### DISCUSSION TITLE
Question:Dkim signing missing in some cases

### DISCUSSION THREAD

**Discussion by c-steve** [rspamd/rspamd#4374] [Q&A]
Hi,
 I try to understand why does my mailserver not sign my emails in a special case, when an email address is present in my whitelisted users list. I m using the multimap module to whitelist some local addresses:

# For accepting whitelisted mail addresses
WHITELIST_SENDER_EMAIL {
      type = "from";
      filter = "email";
      map = "/${LOCAL_CONFDIR}/local.d/whitelist-from.map";
      action = "accept";
      symbol = "SENDER_FROM_WHITELIST_USER"
      description ="Accept emails from this maillist - Action accept";
 }

However , if an email address is not present in my list, dkim signing works fine, and is present.
My dkim config file looks like this:
# If false, messages with empty envelope from are not signed
allow_envfrom_empty = true;

# If true, envelope/header domain mismatch is ignored
allow_hdrfrom_mismatch = false;

# If true, multiple from headers are allowed (but only first is used)
allow_hdrfrom_multiple = false;

# If true, username does not need to contain matching domain
allow_username_mismatch = true;

# Default path to key, can include '$domain' and '$selector' variables
path = "/var/lib/rspamd/dkim/$selector.key";

# Default selector to use
selector = "2022";

# If false, messages from authenticated users are not selected for signing
sign_authenticated = true;

# If false, messages from local networks are not selected for signing
sign_local = true;

# Map file of IP addresses/subnets to consider for signing
# sign_networks = "/some/file"; # or url
# Whether to fallback to global config
try_fallback = true;

# Domain to use for DKIM signing: can be "header" (MIME From), "envelope" (SMTP From) or ">
use_domain = "header";

# Domain to use for DKIM signing when sender is in sign_networks ("header"/"envelope"/"aut>
#use_domain_sign_networks = "header";

# Domain to use for DKIM signing when sender is a local IP ("header"/"envelope"/"auth")
#use_domain_sign_local = "header";

# Whether to normalise domains to eSLD (Second layer domain )
use_esld = false;

# Whether to get keys from Redis
use_redis = false;

# Hash for DKIM keys in Redis
key_prefix = "DKIM_KEYS";

# map of domains -> names of selectors (since rspamd 1.5.3)
#selector_map = "/etc/rspamd/dkim_selectors.map";

# map of domains -> paths to keys (since rspamd 1.5.3)
#path_map = "/etc/rspamd/dkim_paths.map";

# If `true` get pubkey from DNS record and check if it matches private key
check_pubkey = false;
#check_pubkey = true;
# Set to `false` if you want to skip signing if public and private keys mismatch
#allow_pubkey_mismatch = true;
allow_pubkey_mismatch = false;

I User Rspamd V 3.1

Any insight would be appreciated

**Comment by fatalbanana** [comment]
If an action is set on a message most further checks should be skipped.

You might achieve your desired results using `settings` module. Another option is `force_actions` module given `require_action` and/or `honor_action` settings would be used.

The `multimap` config might be fixed by registering a dependency, either in config:
```
require_symbols = "DKIM_SIGNED | !DKIM_SIGNED"
```
or Lua:
```
-- /etc/rspamd/rspamd.local.lua
rspamd_config:register_dependency('WHITELIST_SENDER_EMAIL', 'DKIM_SIGNED')
```

**Reply by c-steve** [reply]
@fatalbanana Unfortunately both of the mentioned workarounds fail. Tried the lua method, and the config method, but for the whitelisted email addresses dkim signing won t be present. Is there any other way ? It s like you said, the moment an action is set on the message, further checks are skipped.
